server/controllers/services.ts

```diff
@@ -48,10 +49,13 @@ function generatePlaylistOEmbed (req: express.Request, res: express.Response) {
 function generateVideoOEmbed (req: express.Request, res: express.Response) {
   const video = res.locals.videoAll
 
+  const regex = /\?[^?]*$/
+  const params = regex.test(req.query.url) ? req.query.url.match(regex) : ''
```

Comment:
Please use https://nodejs.org/api/url.html to extract query parameters

and only reinject params that we know (create a `Set` for example)

Reply:
hello @Chocobozzz,

I need your suggestions here. I have found two approaches for now.

To extract query parameters:
First Approach would be to use WHATWG URL API[#](https://nodejs.org/api/url.html#the-whatwg-url-api) as you suggested:
```
  const paramSearch = new URL(req.query.url).search.split('?').join('').split('=')
  const key = paramSearch[0]
  const { url, ...queryParams } = req.query
  const params = validateParams({ [key]: paramSearch[1], ...queryParams }) || ''
```

The other approach would be to use `decodeQueryString`:
```
  const decode = decodeQueryString(req.query.url)
  const params = validateParams(decode) || ''
```
To reinject params that we know I have created a function similar to `decorateVideoLink` which checks the params validity.

```
function validateParams (options: {
  start?: string
  stop?: string
  loop?: string
  autoplay?: string
  muted?: string
}) {

  const params = new URLSearchParams()

  if (options.start !== undefined && options.start !== null) {
    params.set('start', options.start)
  }

  if (options.stop !== undefined && options.stop !== null) {
    params.set('stop', options.stop)
  }

  if (options.loop !== undefined && options.loop !== null) {
    params.set('loop', options.loop)
  }

  if (options.autoplay !== undefined && options.autoplay !== null) {
    params.set('autoplay', options.autoplay)
  }

  if (options.muted !== undefined && options.muted !== null) {
    params.set('muted', options.muted)
  }

  return buildParams(params)
}

function buildParams (params: URLSearchParams) {
  let hasParams = false
  params.forEach(() => {
    hasParams = true
  })
  if (hasParams) return '?' + params.toString()
}

```
Both the approach works fine but the problem here is using the first approach I get the following result: 
![image](https://user-images.githubusercontent.com/26194365/173758595-53f724f3-c28a-439c-bec5-3a681a346193.png)

And with the second approach I get the following result:
![image](https://user-images.githubusercontent.com/26194365/173758468-ef93b3e7-9b41-4a74-90c1-22d6b37612ab.png)

The first approach does not work when the URL is encoded. 
The second approach does not add the query to the response html.

Reply:
> The other approach would be to use decodeQueryString:

Please use the first approach, the URL object should correctly parse the `url` query param. You can use `searchParams` properties:

```js
new URL(req.query.url).searchParams.forEach((v, k) => {
  console.log(k + '=' + v)
})
```

> To reinject params that we know I have created a function similar to decorateVideoLink which checks the params validity.

I think you can just use `pick` with the parameter names you want to support: https://github.com/Chocobozzz/PeerTube/blob/develop/shared/core-utils/common/object.ts#L1

Reply:
I added a section as you suggested, please review it and let me know what you think!